text/0000-prior-doc-comment.md

```diff
@@ -0,0 +1,94 @@
+- Feature Name: prior_doc_comment
+- Start Date: 2018-03-26
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow `//!` (also called "outer doc comments" below) to be used to annotate the block just prior to them.
+
+# Motivation
+[motivation]: #motivation
+
+This usecase, where the doc comment is put after an enum variant, is likely to be most widely used:
+
+```rust
+enum E {
+    /// doc comments used currently
+    A,
+    B, //! doc comments with this RFC
```

Comment:
It's possible to do this in Haskell with haddock like so:
```Haskell
data Maybe a
  = Nothing -- ^ There is nothing here.
  | Just a  -- ^ We have a measly 'a'.
```

But I'm skeptical that this enhances readability for Rust. At least, if this is possible, I would like it to not be the recommended style by rustfmt.

------------------

This sparked an orthogonal question in my mind... What if we allowed the following:

```rust
fn foo(x: usize,  //! Some comments about x
       y: String, //! Some comments about y
     ) -> Result  //! Comments about return values.
{
    unimplemented!()
}
```

or in Haskell:

```haskell
foo :: Int    -- ^ Some comments about x
    -> String -- ^ Some comments about y
    -> Result -- ^ Comments about return values.
foo x y = undefined
```

I guess using `//!` in this way makes sense in that case.

Reply:
I'm not too keen on the return value. Mind the way this would most likely work is:

```rust
fn foo(x: usize,  //! Some comments about x
       y: String, //! Some comments about y
     ) -> Result  //! Comments about return values.
{ }

fn foo(x: usize,  //! Some comments about x
       y: String, //! Some comments about y
     ) -> Result;  //! Comments about the... hey, wait a second, you
                   //! pesky semicolon! This is actually documenting the
                   //! entire fn item, isn't it?!
```

Reply:
I don't think we would be introducing param documents immediately. Instead, I imagine if this will even be a part of this RFC, it will be just concatenated and we'll leave parameter description boilerplates up to the document author.